diff --git a/feeds_imagegrabber.module b/feeds_imagegrabber.module
index 897c4b8..1d9c157 100644
--- a/feeds_imagegrabber.module
+++ b/feeds_imagegrabber.module
@@ -33,7 +33,7 @@ function feeds_imagegrabber_menu() {
 /**
  * Implements hook_form_BASE_FORM_ID_alter().
  */
-function feeds_imagegrabber_form_node_form_alter(&$form, $form_state) {
+function feeds_imagegrabber_form_node_form_alter(&$form, &$form_state, $form_id) {
   if ($importer_id = feeds_get_importer_id($form['#node']->type)) {
     // Use the values from $form_state if available
     if (isset($form_state['values']['feeds_imagegrabber'])) {
