diff --git a/bean.info b/bean.info
index ce1ea9b..d818d50 100644
--- a/bean.info
+++ b/bean.info
@@ -6,6 +6,7 @@ files[] = includes/bean.info.inc
 files[] = plugins/BeanPlugin.class.php
 files[] = plugins/BeanDefault.class.php
 files[] = includes/translation.handler.bean.inc
+files[] = views/views_handler_argument_bean_type.inc
 files[] = views/views_handler_filter_bean_type.inc
 files[] = views/views_handler_field_bean_type.inc
 files[] = views/views_handler_field_bean_edit_link.inc
diff --git a/views/bean.views.inc b/views/bean.views.inc
index 76d8a2b..f418dc8 100644
--- a/views/bean.views.inc
+++ b/views/bean.views.inc
@@ -7,12 +7,16 @@ function bean_views_data_alter(&$data) {
   $data['bean']['type'] = array(
     'title' => t('Type'),
     'help' => t('The block type'),
+    'argument' => array(
+      'handler' => 'views_handler_argument_bean_type',
+    ),
     'field' => array(
       'handler' => 'views_handler_field_bean_type',
     ),
     'filter' => array(
       'handler' => 'views_handler_filter_bean_type',
     ),
+
   );
 
   $data['bean']['edit_link'] = array(
diff --git a/views/views_handler_argument_bean_type.inc b/views/views_handler_argument_bean_type.inc
new file mode 100644
index 0000000..614f138
--- /dev/null
+++ b/views/views_handler_argument_bean_type.inc
@@ -0,0 +1,15 @@
+<?php
+/**
+ * Argument to accept bean type
+ *
+ */
+class views_handler_argument_bean_type extends views_handler_argument_string {
+  function title() {
+    $title = db_query("SELECT title FROM {bean} v WHERE type = :type", array(':type' => $this->argument))->fetchField();
+
+    if (empty($title)) {
+      return t('No bean type');
+    }
+    return check_plain($title);
+  }
+}
